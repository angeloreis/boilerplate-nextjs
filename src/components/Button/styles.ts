import styled, { css } from 'styled-components'

interface ButtonProps {
  width?: string
  classType: 'primary' | 'success' | 'info' | 'alert' | 'danger' | 'disable'
}

const bgColorButton = {
  primary: '#004e9a',
  success: '#5eb924',
  info: '#428cd4',
  alert: '#fa9806',
  danger: '#ad1212',
  disable: '#dddddd'
}

const bgColorButtonHover = {
  primary: '#0B5394',
  success: '#5eb924',
  info: '#5497d8',
  alert: '#ffb006',
  danger: '#b52929',
  disable: '#cccccc'
}

export const ButtonDefault = styled.button<ButtonProps>`
  min-width: 13.12rem;
  width: ${(props) => props.width};
  padding: 2rem 6rem;
  cursor: pointer;

  ${({ classType }) =>
    classType === 'disable'
      ? css`
          :disabled {
            background: ${bgColorButton[classType]};
            opacity: 0.5;
            cursor: not-allowed;
            border: 1px dashed white;
          }
        `
      : ''}

  background: ${({ classType }) => bgColorButton[classType]};
  color: white;
  border-radius: 5px;
  border: none;
  font-size: 1.2rem;

  transition: 0.2s ease background;

  :hover {
    background: ${({ classType }) => bgColorButtonHover[classType]};
  }
`
