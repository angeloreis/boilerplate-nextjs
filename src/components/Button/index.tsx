import { IButton } from 'types/atomic-design'
import * as S from './styles'

const Button: React.FC<IButton> = ({
  classType = 'primary',
  width = '100%',
  textOfButton,
  onClick
}) => {
  const isDisable = classType === 'disable'
  return (
    <S.ButtonDefault
      classType={classType}
      width={width}
      disabled={isDisable}
      onClick={onClick}
    >
      {textOfButton}
    </S.ButtonDefault>
  )
}

export default Button
