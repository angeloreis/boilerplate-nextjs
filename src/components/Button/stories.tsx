import { Story, Meta } from '@storybook/react'
import Button from '.'

export default {
  title: 'Default button',
  component: Button,
  argTypes: {
    classType: {
      name: 'classType',
      options: ['primary', 'success', 'info', 'alert', 'danger', 'disable'],
      control: { type: 'select' }
    },
    width: {
      name: 'width',
      control: { type: 'text' }
    },
    textOfButton: {
      name: 'text of button',
      control: { type: 'text' }
    }
  }
} as Meta

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const Template = (args: any) => {
  return <Button {...args} />
}

export const Default: Story = Template.bind({})

Default.args = {
  classType: 'primary',
  width: '100%',
  textOfButton: 'Default text'
}
