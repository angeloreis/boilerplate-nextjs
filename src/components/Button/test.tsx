import { render, screen } from '@testing-library/react'

import Button from '.'

describe('Testing default button', () => {
  it('should render a primary button with width 210px and text Text of Test', () => {
    const width = '210px'
    const text = 'Text of Test'
    const type = 'primary'

    const { container } = render(
      <Button classType={type} width={width} textOfButton={text} />
    )

    expect(screen.getByRole('button', { name: /text/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('render an button with alert class and width 210px and text say Alert: test procced', () => {
    const width = '210px'
    const text = 'Alert: test procced'
    const type = 'alert'

    const { container } = render(
      <Button classType={type} width={width} textOfButton={text} />
    )

    const alertButton = screen.getByRole('button', { name: /alert/i })

    expect(alertButton).toBeInTheDocument()
    expect(alertButton).toHaveStyle('background: #fa9806; width: 210px;')
    expect(alertButton).toHaveTextContent('Alert: test procced')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('render an button with disable class and text say Operation suspended', () => {
    const text = 'Operation suspended'
    const type = 'disable'

    const { container } = render(
      <Button classType={type} textOfButton={text} />
    )

    const disableButton = screen.getByRole('button', { name: /suspended/i })

    expect(disableButton).toBeInTheDocument()
    expect(disableButton).toHaveStyle('width: 100%;')
    expect(disableButton).toHaveStyle('background:#dddddd;')
    expect(disableButton).toHaveTextContent('Operation suspended')
    expect(container.firstChild).toMatchSnapshot()
  })
})
