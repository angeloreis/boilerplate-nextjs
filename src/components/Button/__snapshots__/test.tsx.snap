// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `Testing default button render an button with alert class and width 210px and text say Alert: test procced 1`
] = `
.c0 {
  min-width: 13.12rem;
  width: 210px;
  padding: 2rem 6rem;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #fa9806;
  color: white;
  border-radius: 5px;
  border: none;
  font-size: 1.2rem;
  -webkit-transition: 0.2s ease background;
  transition: 0.2s ease background;
}

.c0:hover {
  background: #ffb006;
}

.c0 div {
  padding-left: 8px;
  padding-right: 8px;
}

<button
  class="c0"
  width="210px"
>
  <div>
    <svg
      fill="currentColor"
      height="18"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 192 512"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M176 432c0 44.112-35.888 80-80 80s-80-35.888-80-80 35.888-80 80-80 80 35.888 80 80zM25.26 25.199l13.6 272C39.499 309.972 50.041 320 62.83 320h66.34c12.789 0 23.331-10.028 23.97-22.801l13.6-272C167.425 11.49 156.496 0 142.77 0H49.23C35.504 0 24.575 11.49 25.26 25.199z"
      />
    </svg>
  </div>
  Alert: test procced
</button>
`

exports[
  `Testing default button render an button with disable class and text say Operation suspended 1`
] = `
.c0 {
  min-width: 13.12rem;
  width: 100%;
  padding: 2rem 6rem;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #dddddd;
  color: white;
  border-radius: 5px;
  border: none;
  font-size: 1.2rem;
  -webkit-transition: 0.2s ease background;
  transition: 0.2s ease background;
}

.c0:disabled {
  background: #dddddd;
  opacity: 0.5;
  cursor: not-allowed;
  border: 1px dashed white;
}

.c0:hover {
  background: #cccccc;
}

.c0 div {
  padding-left: 8px;
  padding-right: 8px;
}

<button
  class="c0"
  disabled=""
  width="100%"
>
  <div />
  Operation suspended
</button>
`

exports[
  `Testing default button should render a primary button with width 210px and text Text of Test 1`
] = `
.c0 {
  min-width: 13.12rem;
  width: 210px;
  padding: 2rem 6rem;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #004e9a;
  color: white;
  border-radius: 5px;
  border: none;
  font-size: 1.2rem;
  -webkit-transition: 0.2s ease background;
  transition: 0.2s ease background;
}

.c0:hover {
  background: #0B5394;
}

.c0 div {
  padding-left: 8px;
  padding-right: 8px;
}

<button
  class="c0"
  width="210px"
>
  <div />
  Text of Test
</button>
`
